### ISSUE TITLE
Single rating for one book per one user

### ISSUE THREAD

**Issue by cats-shadow** [bookwyrm-social/bookwyrm#1086]
When I write a review of a book, I rate it in the review. 
However, I can also rate it separately. And they may turn out to be different.
The final rating of a book is the arithmetic average.

It would be logical for one user to give one rating to one book. 
I mean, the database should store only the last book rating (from each user).

**Comment by NonSparkly** [comment]
I ran into this one and decided to double check something.

When going to a book and changing ones rating it, the newly created 'ratings' post doesn't automatically overwrite the users latest rating of that book, meaning that one needs to manually go in and remove the last status post to only have a single rating count.

Combined with the above described arithmetic average, this means that any one user could infinitely rate a re-book and get an outsized amount of influence on that books overall rating.

**Comment by amancior23** [comment]
Hello, I'd love to work on this issue!

**Comment by mouse-reeve** [comment]
It would be to only count the most recent rating

**Comment by RokeJulianLockhart** [comment]
> https://github.com/bookwyrm-social/bookwyrm/issues/1086#issue-892379804

@cats-shadow, I don't think we should only store the last rating — why remove functionality? However, we should disallow mutiple simultaneous reviews per user. I think that past review should only be exposed a rating history, if that makes sense. I have found the rating system a little confusing too.

**Comment by cats-shadow** [comment]
@RokeJulianLockhart 
For myself, I don't see the point of ranking history, to be honest.
I am absolutely not interested in how my attitude to the book changed, in addition I remember only the last impressions. :)
Rating history makes sense only if you regularly reread the book (not once and not twice), and keep some kind of analytics. Personally, I don't do that. I just read books. The good ones (rating 4 or 5 on my scale) I sometimes reread. The ones rated 3 and below on my scale, I reread only by accident, only if I forget that I have read them (so far, I have no complaints about my memory :) ).

But the ability to rate a book in several unrelated places of the interface just leads to the described.

See screenshot.
![изображение](https://github.com/bookwyrm-social/bookwyrm/assets/16667173/e5bebca5-0855-48b8-b85f-29a82726a866)

**Comment by amancior23** [comment]
Hello @mouse-reeve, @Robel-003 and I have spent some time navigating through various directories, with a focus on the model and view layers, attempting to familiarize ourselves with the functionalities encapsulated within each file. Despite these efforts, we’ve hit a roadblock in pinpointing the exact implementation of the rating system. We were wondering if you could provide some guidance or point us in the direction of where the rating system's logic is primarily handled within the codebase. Any insights or suggestions on specific files or modules to look at would be greatly appreciated. Thank you!

**Comment by joachimesque** [comment]
The calculations for books are made for each use of the calculation. Right now there’s no common code, it’s all replicated at various places:

- https://github.com/bookwyrm-social/bookwyrm/blob/7469f1f4ca0a411a823f3e2a5c24655985bdcd60/bookwyrm/preview_images.py#L460
- https://github.com/bookwyrm-social/bookwyrm/blob/7469f1f4ca0a411a823f3e2a5c24655985bdcd60/bookwyrm/views/books/books.py#L90
- https://github.com/bookwyrm-social/bookwyrm/blob/7469f1f4ca0a411a823f3e2a5c24655985bdcd60/bookwyrm/templatetags/rating_tags.py#L23
- https://github.com/bookwyrm-social/bookwyrm/blob/7469f1f4ca0a411a823f3e2a5c24655985bdcd60/bookwyrm/views/annual_summary.py#L80

I haven’t contributed in a long while so I might be mistaken in my approach, but here’s how I would go about it: the ratings depend on the book reviews. Depending on the different contexts, we have different filters for the `models.Review.objects`. All of these queries should have an additionnal filter that only takes one review _with rating_ per user, and that that review should be the latest. It might be a good idea to see if you can make an utility to get the rating of a book, so that these four occurences could use it as an utility, instead of reimplementing the same calculation based on query filters that are slightly different.
